cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SourceControlManagementHttpHandler.java

```diff
@@ -104,6 +123,85 @@ public void deleteRepository(FullHttpRequest request, HttpResponder responder,
                                                               namespaceId));
   }
 
+  /**
+   * Pushes a set of applications configs to linked repository in Json format. It expects a post body that has an
+   * optional commit message and an array of objects, with each object specifying the application and version. The
+   * size of applications list should not be more than 10.
+   * E.g.
+   *
+   * <pre>
+   * {@code
+   * {
+   *   "apps": [
+   *   {"application":"XYZ", "version": "1.23"},
+   *   {"application":"ABC", "version": "1.2.4"},
+   *   {"application":"FOO", "version": "1.23.4"}
+   *   ],
+   *   "commitMessage": "pushed apps XYZ, ABC and FOO"
+   * }
+   * }
+   *
+   * </pre>
+   * The response will be an array of {@link BatchApplicationPushResult} object, which either indicates a success (200)
+   * or failure for each of the requested application in the same order as the request.
+   */
+  @POST
+  @Path("/push")
+  public void pushApps(FullHttpRequest request, HttpResponder responder,
+                       @PathParam("namespace-id") String namespaceId) throws Exception {
+    checkSourceControlFeatureFlag();
+    NamespaceId namespace = validateNamespaceId(namespaceId);
+    ImmutablePair<List<ApplicationId>, String> appsRequest = validateAndGetAppsRequest(request, namespace);
+
+    List<ApplicationId> appIds = appsRequest.getFirst();
+
+    Map<ApplicationId, SourceControlMeta> sourceControlMetas =
+      sourceControlService.pushApps(namespace, appIds, appsRequest.getSecond());
+    
+    List<BatchApplicationPushResult> result = new ArrayList<>();
+    for (ApplicationId appId : appIds) {
+      SourceControlMeta sourceControlMeta = sourceControlMetas.get(appId);
+      if (sourceControlMeta == null) {
+        result.add(new BatchApplicationPushResult(appId, new NotFoundException(appId)));
+      } else {
+        result.add(new BatchApplicationPushResult(appId, sourceControlMeta));
+      }
+    }
+
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(result));
+  }
+
+  private ImmutablePair<List<ApplicationId>, String> validateAndGetAppsRequest(
+    FullHttpRequest request, NamespaceId namespace) throws BadRequestException {
+    
+    PushAppsRequest appsRequest;
+    try {
+      appsRequest = parseBody(request, PushAppsRequest.class);
+    } catch (JsonSyntaxException e) {
+      throw new BadRequestException("Invalid request body: " + e.getMessage());
+    }
+    
+    if (appsRequest == null || appsRequest.getApps() == null || appsRequest.getApps().isEmpty()) {
+      throw new BadRequestException("Please specify a list of applications.");
+    }
+
+    List<ApplicationId> appIds = new ArrayList<>();
+
+    for (PushAppsRequest.App app : appsRequest.getApps()) {
+      try {
+        appIds.add(namespace.app(app.getName(), app.getVersion()));
+      } catch (Exception e) {
+        throw new BadRequestException("Invalid application name or version", e);
+      }
+    }
+
+    if (appIds.size() > 10) {
```

Comment:
Maybe we should have this limit configurable through cdap-site.xml. In case we find it needs to be changed depending on the CPU/memory available to CDAP.

Reply:
Have put this in the cdap-default.xml. Users are able to change it in cdap-site.xml and override the default one